File: src/pbl/consts/weight.nim
# Values taken from:
# https://pepcalc.com/notes.php?mw
import amino

const NTermH*  =  1.00797
const CTermOH* = 17.00738

const MolecularWeight* = [
    aaA: 71.07793,
    aaC: 103.1454,
    aaD: 115.0873,
    aaE: 129.1139,
    aaF: 147.1734,
    aaG: 57.05138,
    aaH: 137.1394,
    aaI: 113.1576,
    aaK: 128.1724,
    aaL: 113.1576,
    aaM: 131.1985,
    aaN: 114.1028,
    aaP: 97.11508,
    aaQ: 128.1293,
    aaR: 156.1861,
    aaS: 87.07733,
    aaT: 101.1039,
    aaV: 99.13103,
    aaW: 186.2095,
    aaY: 163.1728
]

converter toMolecularWeight*(c: char): float =
  result = case c:
  of 'A': 71.07793
  of 'C': 103.1454
  of 'D': 115.0873
  of 'E': 129.1139
  of 'F': 147.1734
  of 'G': 57.05138
  of 'H': 137.1394
  of 'I': 113.1576
  of 'K': 128.1724
  of 'L': 113.1576
  of 'M': 131.1985
  of 'N': 114.1028
  of 'P': 97.11508
  of 'Q': 128.1293
  of 'R': 156.1861
  of 'S': 87.07733
  of 'T': 101.1039
  of 'V': 99.13103
  of 'W': 186.2095
  of 'Y': 163.1728
  of '_': 0.0  
  else: 
    raise newException(ValueError, "invalid character")

File: framework/src/consts/weight.rs
//! Values taken from:
//! https://pepcalc.com/notes.php?mw

/// in [Da]
#[non_exhaustive]
pub struct Weight;

impl Weight {
    pub const A: f32 = 71.07793;
    pub const C: f32 = 103.1454;
    pub const D: f32 = 115.0873;
    pub const E: f32 = 129.1139;
    pub const F: f32 = 147.1734;
    pub const G: f32 = 57.05138;
    pub const H: f32 = 137.1394;
    pub const I: f32 = 113.1576;
    pub const K: f32 = 128.1724;
    pub const L: f32 = 113.1576;
    pub const M: f32 = 131.1985;
    pub const N: f32 = 114.1028;
    pub const P: f32 = 97.11508;
    pub const Q: f32 = 128.1293;
    pub const R: f32 = 156.1861;
    pub const S: f32 = 87.07733;
    pub const T: f32 = 101.1039;
    pub const V: f32 = 99.13103;
    pub const W: f32 = 186.2095;
    pub const Y: f32 = 163.1728;

    pub const NTERM_H: f32 = 1.00797;
    pub const CTERM_OH: f32 = 17.00738;

    pub fn convert(c: char) -> f32 {
        match c {
            'A' => Weight::A,
            'C' => Weight::C,
            'D' => Weight::D,
            'E' => Weight::E,
            'F' => Weight::F,
            'G' => Weight::G,
            'H' => Weight::H,
            'I' => Weight::I,
            'K' => Weight::K,
            'L' => Weight::L,
            'M' => Weight::M,
            'N' => Weight::N,
            'P' => Weight::P,
            'Q' => Weight::Q,
            'R' => Weight::R,
            'S' => Weight::S,
            'T' => Weight::T,
            'V' => Weight::V,
            'W' => Weight::W,
            'Y' => Weight::Y,
            _ => unreachable!(),
        }
    }
}
